src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+            : STTx::RequireFullyCanonicalSig::no;
+        auto const sigResult =
+            ctx.tx.checkBatchSign(requireCanonicalSig, ctx.rules);
+        if (!sigResult)
+        {
+            JLOG(ctx.j.warn()) << "Batch: invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+Batch::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.rules().enabled(featureBatch))
+        return temDISABLED;
+
+    return tesSUCCESS;
+}
+
+TER
+Batch::doApply()
+{
+    Sandbox sb(&ctx_.view());
+    bool changed = false;
+    auto const flags = ctx_.tx.getFlags();
+
+    AccountID const outerAccount = ctx_.tx.getAccountID(sfAccount);
+
+    TER result = tesSUCCESS;
+    ApplyViewImpl& avi = dynamic_cast<ApplyViewImpl&>(ctx_.view());
+    OpenView subView(&sb);
+
+    auto const& txns = ctx_.tx.getFieldArray(sfRawTransactions);
+    bool const not3rdParty = std::any_of(
+        txns.begin(), txns.end(), [outerAccount](STObject const& txn) {
+            return txn.getAccountID(sfAccount) == outerAccount;
+        });
+
+    for (STObject txn : txns)
+    {
+        OpenView innerView(&subView);
+
+        STTx const stx = STTx{std::move(txn)};
+        auto const [ter, applied] =
+            ripple::apply(ctx_.app, innerView, stx, tapFAIL_HARD, ctx_.journal);
+
+        if (applied)
+            innerView.apply(subView);
+
+        changed = true;
+
+        // Add Inner Txn Metadata
+        STObject meta{sfBatchExecution};
+        std::string res = transToken(ter);
+        meta.setFieldVL(sfInnerResult, ripple::Slice{res.data(), res.size()});
+        meta.setFieldU16(sfTransactionType, stx.getTxnType());
+        if (ter == tesSUCCESS || isTecClaim(ter))
+            meta.setFieldH256(sfTransactionHash, stx.getTransactionID());
+        avi.addBatchExecutionMetaData(std::move(meta));
+
+        if (ter != tesSUCCESS)
+        {
+            // Atomic Revert on non tec failure
+            if (!isTecClaim(ter))
+            {
+                JLOG(ctx_.journal.warn()) << "Batch: Inner txn failed." << ter;
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+
+            if (flags & tfUntilFailure)
+            {
+                result = tesSUCCESS;
+                break;
+            }
+            if (flags & tfOnlyOne)
+            {
+                continue;
+            }
+            if (flags & tfAllOrNothing)
+            {
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+        }
+
+        if (ter == tesSUCCESS && flags & tfOnlyOne)
+        {
+            result = tesSUCCESS;
+            break;
+        }
+    }
+
+    // Apply SubView & PreviousFields
+    if (changed)
+    {
+        // Only required when not 3rd Party
+        if (not3rdParty)
+            ctx_.batchPrevious(avi);
+
+        ctx_.applyOpenView(subView);
```

Comment:
what's the reason for this call here?

Reply:
`batchPrevious` applies the inner transactions to the open view. Before this, they are not actually applied, or they are applied to the subview but not the BatchOuter OpenView. 

`batchPrevious` is a workaround for the "previousFields issue". The issue is thatwhen the outer batch txn is applied at the end, it applies the old/stale view (from before the inner txns). In the case where the outer batch account has an inner batch txn that adjusts its account root, the inner adjustments are not actually applied. They are on the open view but not applied. So this function applies the open view onto the base view and handles this issue. 

This should be thoroughly reviewed. There might be another way. It might be an issue with how the views are applied.

Reply:
Something still doesn't feel right here. We are only applying the changes if the ledger is closed, and never applied when the ledger is open.

When working with  unit tests, after I take out `env.close() `after the the submission of a batch tx, all subsequent validations fail because they were never applied to the open ledger.

Unfortunately I don't have a clear answer to this. Perhaps @ximinez would know if this is the right way to do it.

Reply:
Makes me thinking, if we change the implementation of this function to simply do without the if-condition:
```
void
ApplyContext::applyOpenView(OpenView& open)
{
    open.apply(base_);
}
```

This would unconditionally apply the states of the inner txn to the `base_`, which _might_ avoid the need to have another `batchPrevAcctRootFields_` to store the previous state of account root, and metadata generation might work by themselves.

Obviously I haven't tested my hypothesis and you might have already tried this approach (which might not work). But I would just like to have a bit more understanding on the decisions here (and why it doesn't work)